src/Abp/Domain/Repositories/RepositoryExtensions.cs

```diff
@@ -77,5 +80,44 @@ public static class RepositoryExtensions
 
             throw new ArgumentException($"Given {nameof(repository)} is not inherited from {typeof(AbpRepositoryBase<TEntity, TPrimaryKey>).AssemblyQualifiedName}");
         }
+
+        public static async Task HardDelete<TEntity, TPrimaryKey>(this IRepository<TEntity, TPrimaryKey> repository, TEntity entity)
+            where TEntity : class, IEntity<TPrimaryKey>, ISoftDelete
+        {
+            var repo = ProxyHelper.UnProxy(repository) as IRepository<TEntity, TPrimaryKey>;
+            if (repo == null)
+            {
+                throw new ArgumentException($"Given {nameof(repository)} is not inherited from {typeof(IRepository<TEntity, TPrimaryKey>).AssemblyQualifiedName}");
+            }
+
+            var extensionData = ((IUnitOfWorkManagerAccessor)repo).UnitOfWorkManager.Current.ExtensionData;
+            var hardDeleteEntities = extensionData.GetOrAdd(UnitOfWorkExtensionDataTypes.HardDelete, () => new HashSet<string>()) as HashSet<string>;
+
+            var tenantId = GetCurrentTenantIdOrNull(repo.GetIocResolver());
+            var hardDeleteKey = EntityHelper.GetHardDeleteKey(entity, tenantId);
+
+            if (hardDeleteEntities == null)
+            {
+                throw new ArgumentNullException(nameof(hardDeleteEntities));
```

Comment:
hardDeleteEntities is not an argument of this method. Throw another type of exception normally. However, this should not be needed since extensionData.GetOrAdd can not return null. So, completely remove this check.